tests/FsCheck.Test/Gen.fs

```diff
@@ -181,6 +182,29 @@ module Gen =
             && (Array2D.length2 arr <= cols) 
             && (seq { for elem in arr do yield elem :?> int} |> Seq.forall ((=) v))
 
+
+    type MaybeNull =
+        {
+            SomeField : int
+        }
+
+    [<CLIMutable>]
+    type MaybeHaveNull =
+        {
+             MightBeNull : MaybeNull
+        }
+
+    [<Fact>]
+    let ``should support null as a shrunken value even without AllowNullLiteral (for C# compatibility)``() =
+        let der = Arb.Default.Derive()
+        let mhn = { MightBeNull = { SomeField = 5 } }
+        let prop = mhn.GetType().GetProperty("MightBeNull", BindingFlags.Public ||| BindingFlags.Instance)
+        prop.SetValue(mhn, null)
+        // ensure that no exception occurs
+        der.Shrinker mhn
+          |> Seq.length
+          |> ignore
```

Comment:
One might be able to make this test a bit simpler, but unfortunately the example you provided won't quite do it.  The null value is inferred to have the 'object' type, so reflectShrinkObj does not recognize a record and therefore does not end up calling shrinkChildren.  shrinkChildren is what calls getShrink, which is what calls to ShrinkerObj (where I made the code change).  
Moreover, if you send in a record with an (illegitimate) null value directly to the derive shrinker, it will crash in a different way than I was seeing (the outer loop in shrinkChildren's seq generator).  I'm not sure whether that case should be considered a problem, since one generally does not invoke the shrinker directly, and sending it a null value that is recognized as something other than 'object' requires some deliberate misbehavior.  My test case attempted to replicate my problem as closely as possible, but in F# instead of the C# where I was seeing it.  But let me know if you still want me to try to find a simpler test case.